feat: support Tron

ulissesferreira opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Add support for the Tron network through a preinstalled Tron Wallet Snap

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/35984?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: Added support for Tron

## **Related issues**

n/a

## **Manual testing steps**

1. Checkout this branch locally
2. Disable BIP44 changes using the `enableMultichainAccountsState2` flag on the `.manifest-overrides.json`
3. After onboarding, "Create Tron account"

## **Screenshots/Recordings**

### **Before**

n/a

### **After**

<img width="1547" height="859" alt="Screenshot 2025-09-30 at 18 44 50" src="https://github.com/user-attachments/assets/2e7439b5-d508-400d-8ad9-f05db9519396" />


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds Tron network/account support powered by a preinstalled Tron Wallet Snap, with address validation, UI flows, controllers, build flags, and policies.
> 
> - **Multichain/Networks & Assets**
>   - Introduces Tron networks (`TrxScope.{Mainnet,Nile,Shasta}`) with nicknames, icons, block explorers, and native asset mappings (`TRX`, CAIP19) in `shared/constants/multichain/{networks,assets}.ts` and `shared/constants/network.ts`.
>   - Extends network ordering and enablement to include Tron (`TrxScope.Mainnet`).
> - **Snaps & Build**
>   - Preinstalls `@metamask/tron-wallet-snap` in `shared/lib/snaps/snaps.ts`, `app/scripts/constants/snaps.ts`, and `builds.yml` (feature flag `tron`, assets copy rule).
>   - Exposes Tron Snap IDs/names via `shared/lib/accounts/tron-wallet-snap.ts`; whitelists in `shared/lib/accounts/snaps.ts`.
> - **Accounts/Controllers**
>   - Adds Tron provider to `MultichainAccountService` with enable/disable via remote feature flags (`tronAccounts`) in `app/scripts/controller-init/multichain/...`.
>   - Updates network enablement init to include Tron and auto-enable EVM mainnet when only a Tron/SOL/BTC network is enabled.
>   - `metamask-controller`: counts/discovers/creates Tron accounts; integrates Tron Snap client and types; updates tests accordingly.
> - **Validation & Utilities**
>   - Implements Tron address validation (Base58/hex) in `shared/lib/multichain/addresses/tron.ts` with tests; adds `isTronAddress`, `isTronChainId` and recipient validations.
>   - Maps Tron account types to snaps/names in `shared/constants/multichain/accounts.ts`.
> - **UI**
>   - Adds Tron to account creation flows, menus, wallet details, account overview/details (`ui/components/...`, `ui/pages/...`); supports Tron in send flow (type detection, recipients, validation, loader).
>   - Network manager shows Tron where applicable.
> - **Selectors**
>   - Extends selectors for Tron enablement, detection, balances, last-selected account, and multichain configs (`ui/selectors/...`).
> - **Localization & Assets**
>   - Adds `networkNameTron` and `addNewTronAccountLabel` strings (`app/_locales/*`); adds `app/images/tron-logo.svg`.
> - **Policies/Deps**
>   - Adds dependency `@metamask/tron-wallet-snap` and `@scure/base` bump; updates LavaMoat policies to reference `@scure/base` directly.
>   - Updates `.depcheckrc.yml`, `.github/CODEOWNERS` (new networks path).
> - **Tests**
>   - Adds unit tests for Tron address validation and multichain selectors; updates controller tests for discovery counts.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f1c81bda01e6059202ce57e5a89bb6388de30f73. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ulissesferreira committed:
feat: support Tron

ngalee8311 commented:
feat/support-tron
files changed

aganglada commented on app/scripts/lib/snap-keyring/keyring-snaps-permissions.ts:
```diff
@@ -55,8 +68,20 @@ const WEBSITE_ALLOWED_METHODS: string[] = [
+  KeyringRpcMethod.RejectRequest,
+  KeyringRpcMethod.ListAccountTransactions,
+  KeyringRpcMethod.ListAccountAssets,
+  KeyringRpcMethod.DiscoverAccounts,
```
same here

aganglada commented on shared/constants/multichain/assets.ts:
```diff
@@ -42,4 +52,7 @@ export const MULTICHAIN_NETWORK_TO_ASSET_TYPES: Record<
   [MultichainNetworks.BITCOIN_SIGNET]: [MultichainNativeAssets.BITCOIN_SIGNET],
+  [MultichainNetworks.TRON]: [MultichainNativeAssets.TRON],
+  [MultichainNetworks.TRON_NILE]: [MultichainNativeAssets.TRON_NILE],
+  [MultichainNetworks.TRON_SHASTA]: [MultichainNativeAssets.TRON_SHASTA],
```
Do we need codefence here? or maybe it's only a map

aganglada commented on shared/constants/multichain/networks.ts:
```diff
@@ -290,6 +335,72 @@ export const MULTICHAIN_PROVIDER_CONFIGS: Record<
+  /**
+   * Tron
+   */
+  [MultichainNetworks.TRON]: {
```
same here, in this file I dont see codefences, but maybe it's ok

aganglada commented on ui/hooks/accounts/useMultichainWalletSnapClient.ts:
```diff
@@ -20,6 +20,10 @@ import {
+import {
+  TRON_WALLET_NAME,
+  TRON_WALLET_SNAP_ID,
+} from '../../../shared/lib/accounts/tron-wallet-snap';
```
no codefences

ulissesferreira replied:
I didn't see one for Bitcoin so I didn't add it. But I can definitely add it, you think it's a good idea?

aganglada replied:
lets add it I think

ulissesferreira commented on ui/components/multichain/account-menu/account-menu.tsx:
```diff
@@ -132,6 +139,10 @@ export const SNAP_CLIENT_CONFIG_MAP: Record<
+  [ACTION_MODES.ADD_TRON]: {
+    clientType: WalletClientType.Tron,
+    chainId: MultichainNetworks.TRON,
+  },
```
For things like constants we don't need condefences, that is why Bitcoin doesn't have them either. I added them whenever Bitcoin has them too, to match the existing logic

ulissesferreira commented on ui/components/multichain/network-manager/components/default-networks/default-networks.tsx:
```diff
@@ -194,6 +202,13 @@ const DefaultNetworks = memo(() => {
           }
+          if (
+            trxAccountGroup &&
+            isFlask() &&
```
To match the Bitcoin implementation too

<img width="355" height="307" alt="Screenshot 2025-10-20 at 11 29 16" src="https://github.com/user-attachments/assets/a9662fd8-0bb5-4272-b082-f9dfe0d9eeb0" />

OGPoyraz committed:
Temp Tron update in send (#36987)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/36987?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the
CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing
description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and
accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Enables Tron in the send flow with chain detection, account filtering,
address validation, and routing for pending approvals.
>
> - **Confirmations · Send**:
> - Add Tron chain detection via `isTronChainId` and expose
`isTronSendType` in `useSendType`.
> - Include Tron accounts in recipients using `isTronAccountForSend` and
`isTronSendType` in `useAccountRecipients`.
> - Validate Tron recipients with `validateTronAddress` and
`isTronAddress` in `useRecipientValidation`.
> - Route pending Tron send approvals in `Loader` using
`TRON_WALLET_SNAP_ID`.
> - **Utils**:
>   - `network.ts`: add `isTronChainId`.
>   - `account.ts`: add `isTronAccountForSend`.
>   - `sendValidations.ts`: add `validateTronAddress`.
>
> <sup>Written by [Cursor
Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit
52980e9db7bf403af64f7a4497c8e8773252133a. This will update automatically
on new commits. Configure
[here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Co-authored-by: Ulisses Ferreira <ulisses@hey.com>

ulissesferreira commented on ui/components/multichain/network-manager/components/default-networks/default-networks.tsx:
```diff
@@ -103,6 +103,14 @@ const DefaultNetworks = memo(() => {
 
+  let trxAccountGroup = null;
+
+  if (isFlask()) {
```
I am guessing it is removed before we ship no?

aganglada replied:
No idea, I think worth getting @MetaMask/accounts-engineers eyes here

ulissesferreira replied:
This has been removed on the latest `main` so it's no longer a problem

OGPoyraz reviewed (dismissed):
Confirmation changes LGTM - tested this in local

ccharly commented on ui/components/multichain/network-manager/network-manager.test.tsx:
```diff
@@ -9,6 +9,12 @@ import mockState from '../../../../test/data/mock-state.json';
+  value: false,
+  writable: true,
+});
+
```
Is this really required? This would """enable""" flask for all existing tests and could have some side-effects? I don't recall seeing any `isFlask` check for this PR?

ccharly commented on ui/pages/confirmations/hooks/send/useSendType.ts:
```diff
@@ -24,18 +25,25 @@ export const useSendType = () => {
         (isBitcoinSendType && assetIsNative),
       isSolanaSendType,
-      isBitcoinSendType,
+      isTronSendType,
```
Seems to have been fixed, we can resolve that?

ccharly commented on ui/pages/confirmations/utils/network.ts:
```diff
@@ -7,3 +9,14 @@ export function getImageForChainId(chainId: string): string | undefined {
+      TrxScope.Shasta.toString(),
+    ].includes(chainId);
+  }
+  return false;
```
Nit: Cosmetic: This could have been written using 1 single-expression

```suggestion
  return isCaipChainId(chainId) &&
    [
      TrxScope.Mainnet.toString(),
      TrxScope.Nile.toString(),
      TrxScope.Shasta.toString(),
    ].includes(chainId);
```

ccharly commented on ui/pages/multichain-accounts/account-details/account-type-utils.ts:
```diff
@@ -131,3 +138,12 @@ export const isInstitutionalEVMAccount = (
 };
+
+/**
+ * Checks if an account is a Bitcoin account
```
```suggestion
 * Checks if an account is a Tron account
```

ccharly commented on ui/selectors/multichain.ts:
```diff
@@ -671,3 +721,13 @@ export function getLastSelectedSolanaAccount(state: MultichainState) {
+      (a, b) => (b.metadata.lastSelected ?? 0) - (a.metadata.lastSelected ?? 0),
+    );
+  return sortedNonEvmAccounts.length > 0 ? sortedNonEvmAccounts[0] : undefined;
+}
```
Note for other reviewers: This function is added since we have similar functions Bitcoin and Solana, but with the new account model (BIP-44) we should not rely on the `AccountsController` internals anymore. This kind of logic should use account grounds and not use "single accounts" anymore.

I'll make a note to remove/cleanup those functions later on, for now we can keep it (like the others)

ulissesferreira commented on app/scripts/metamask-controller.test.js:
```diff
@@ -3664,6 +3666,7 @@ describe('MetaMaskController', () => {
         // - 3 discovered Bitcoin accounts
         // - 1 Solana default account
+        // - 1 Tron default account
         expect(mockCreateAccount).toHaveBeenCalledTimes(3 + 1);
```
Absolutely! I had added the +1 and the comment before realizing that I did not change the discovery for this particular test. The comment was a leftover that should be removed. Done. Great catch!

ulissesferreira commented on shared/constants/network.ts:
```diff
@@ -1128,6 +1129,7 @@ export const CHAIN_ID_TO_NETWORK_IMAGE_URL_MAP: Record<string, string> = {
   [MultichainNetworks.BITCOIN]: BITCOIN_IMAGE_URL,
   [MultichainNetworks.BITCOIN_TESTNET]: BITCOIN_TESTNET_IMAGE_URL,
   [MultichainNetworks.BITCOIN_SIGNET]: BITCOIN_SIGNET_IMAGE_URL,
+  [MultichainNetworks.TRON]: TRON_IMAGE_URL,
```
(ChatGPT voice) "You are absolutely right!". The thing is I was wondering why we didn't do that for Solana here though 🤔. I was thinking that maybe Solana and Tron had more in comment here, where we didn't really need different logos... Branding wise I see they don't use different logos. Maybe I should just remove the extra logos and point all of them to the same. What do you think?

ulissesferreira replied:
(did this preemptively but curious about your thoughts)

ulissesferreira committed:
chore: improvements suggested in PR comments

ulissesferreira commented on ui/selectors/selectors.js:
```diff
@@ -3191,6 +3202,17 @@ export function getIsSolanaSupportEnabled(state) {
+ */
+export function getIsTronSupportEnabled(state) {
+  const { addTronAccount } = getRemoteFeatureFlags(state);
+  return Boolean(addTronAccount);
```
Changed this + updated the feature flag on the appropriate platform 👍

ulissesferreira committed:
chore: update lavamoat policies

ulissesferreira commented:
@MetaMask/policy-reviewers @MetaMask/supply-chain I ran `yarn lavamoat:auto` and I commited some changes here that are making 2 CI steps fail. Can I have your 👀 on those changes?

metamaskbot committed:
Update LavaMoat policies

ulissesferreira committed:
fix: Tron decimals in networks constant

MajorLift commented on lavamoat/browserify/main/policy.json:
```diff
@@ -2953,6 +2967,36 @@
+        "location.href": true,
+        "navigator": true,
+        "queueMicrotask": true,
+        "setTimeout": true
```
@naugtur Would you be able to confirm whether these policy entries are safe? There's also a `"fetch": true` entry under `tronweb>ethers`.

aganglada replied:
`tronweb` is the official tron sdk, this let us read and execute on the tron blockchain.
Not sure if it's enough information, but let me know otherwise.

aganglada committed:
chore: fix lint

aganglada committed:
chore: importscripts to false

aganglada committed:
chore: replace tronweb with an address validator

aganglada committed:
fix: lint

aganglada committed:
fix: lint

aganglada committed:
fix: lint

OGPoyraz reviewed (dismissed):
Confirmation changes LGTM

aganglada committed:
chore: add tron to main

aganglada committed:
chore: fix tron addres validator

aganglada commented on shared/lib/multichain/addresses/tron.ts:
```diff
@@ -0,0 +1,281 @@
+ */
+
+import { is, string } from '@metamask/superstruct';
+import { sha256 } from 'js-sha256';
```
For @MetaMask/policy-reviewers this is introduced to keep the tron address validator syncronous. In the past `tronweb` was using the same library to validate addresses. But he dependency was to big and introducing many other duplicate dependencies with it.

metamaskbot committed:
Update LavaMoat policies

aganglada committed:
chore: fix tests

aganglada committed:
chore: simplify tron validation

aganglada committed:
chore: rollback policy

aganglada committed:
chore: rollback policy

aganglada commented:
@metamaskbot update-policies

ulissesferreira committed:
chore: bump Tron snap to version 1.7.1

aganglada commented on ui/selectors/multichain/networks.ts:
```diff
@@ -103,12 +107,15 @@ export const getIsNonEvmNetworksEnabled = createDeepEqualSelector(
         solanaEnabled = true;
       }
-      if (bitcoinEnabled && solanaEnabled) {
+      if (scopes?.includes(TrxScope.Mainnet)) {
```
the logic is bellow, just checked

metamaskbot committed:
Update LavaMoat policies

FrederikBolding commented on yarn.lock:
```diff
@@ -10338,6 +10345,13 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@scure/base@npm:^1.2.5":
```
Can be deduped with `yarn dedupe`

aganglada committed:
chore: dedupe

aganglada commented on development/build/scripts.js:
```diff
@@ -1017,6 +1017,7 @@ function setupBundlerDefaults(
             './**/node_modules/marked',
             './**/node_modules/@solana',
             './**/node_modules/axios',
+            './**/node_modules/tronweb',
```
Thanks!

aganglada commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

aganglada committed:
chore: fix cursor feedback

FrederikBolding reviewed (dismissed):
Policy and Snaps changes LGTM

aganglada committed:
chore: fix builds

gantunesr commented on ui/pages/multichain-accounts/account-details/multichain-account-details.tsx:
```diff
@@ -62,6 +63,11 @@ export const MultichainAccountDetails = () => {
+      case 'tron':
+        return (
+          <TronAccountDetails address={address as string} account={account} />
+        );
```
Again, this should be unnecessary with BIP-44 UI

gantunesr reviewed (approved):
Will approve this PR but left comments about the overall tech debt this will add after the release of BIP-44

aganglada commented on shared/lib/multichain/addresses/tron.ts:
```diff
@@ -0,0 +1,122 @@
+ *
+ * This utility validates Tron addresses in both Base58 and Hex formats.
+ * Uses @noble/hashes for SHA256 and @scure/base for Base58 encoding.
+ */
```
Mobile uses `tronweb` for now which is basically a version of this simplified code. We will be doing that on mobile next

aganglada commented on shared/constants/multichain/accounts.ts:
```diff
@@ -1,6 +1,15 @@
+  SolAccountType,
+  ///: BEGIN:ONLY_INCLUDE_IF(tron)
+  TrxAccountType,
+  ///: END:ONLY_INCLUDE_IF
```
We will remove all code fences once tron is on rc, Bitcoin is already on RC

Gudahtt reviewed (approved):
Codeowner changes LGTM, I'm approving to satisfy the codeowner requirement for that file. I didn't review the rest
